File: src/main/java/kuvaldis/algorithm/sort/InsertionSort.java
package kuvaldis.algorithm.sort;

import java.util.stream.IntStream;

public class InsertionSort extends Sort {

    @Override
    public int[] sort(final int[] input) {
        final int[] output = new int[input.length];
        System.arraycopy(input, 0, output, 0, output.length);
        IntStream.range(1, output.length).forEach(i -> {
            int key = output[i];
            int j = i;
            while (j > 0 && output[j - 1] > key) {
                output[j] = output[j - 1];
                j--;
            }
            output[j] = key;
        });
        return output;
    }
}

File: python/1_1/insertion_sort.py
def sort(a):
    for i in range(1, len(a)):
        val = a[i]
        j = i
        while j > 0 and a[j - 1] > val:
            a[j] = a[j - 1]
            j -= 1
        a[j] = val


def invert_sort(a):
    for i in range(1, len(a)):
        val = a[i]
        j = i
        while j > 0 and a[j - 1] < val:
            a[j] = a[j - 1]
            j -= 1
        a[j] = val


a = [1, 4, 2, 5, 1, 3, 8, 11, 2, 6, 1]
sort(a)
print(a)
a = [1, 4, 2, 5, 1, 3, 8, 11, 2, 6, 1]
invert_sort(a)
print(a)
